File: app/src/main/java/dev/dimension/flare/ui/screen/profile/Bluesky.kt
package dev.dimension.flare.ui.screen.profile

import androidx.compose.animation.AnimatedVisibilityScope
import androidx.compose.animation.ExperimentalSharedTransitionApi
import androidx.compose.animation.SharedTransitionScope
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.FilledTonalButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import com.eygraber.compose.placeholder.material3.placeholder
import dev.dimension.flare.R
import dev.dimension.flare.ui.component.HtmlText2
import dev.dimension.flare.ui.component.MatricesDisplay
import dev.dimension.flare.ui.model.UiRelation
import dev.dimension.flare.ui.model.UiState
import dev.dimension.flare.ui.model.UiUser
import dev.dimension.flare.ui.model.descriptionDirection
import dev.dimension.flare.ui.model.onSuccess
import dev.dimension.flare.ui.theme.screenHorizontalPadding
import kotlinx.collections.immutable.persistentMapOf

context(AnimatedVisibilityScope, SharedTransitionScope)
@OptIn(ExperimentalSharedTransitionApi::class)
@Composable
internal fun BlueskyProfileHeader(
    user: UiUser.Bluesky,
    relationState: UiState<UiRelation>,
    onFollowClick: (UiRelation.Bluesky) -> Unit,
    onAvatarClick: () -> Unit,
    onBannerClick: () -> Unit,
    isMe: UiState<Boolean>,
    menu: @Composable RowScope.() -> Unit,
    modifier: Modifier = Modifier,
    expandMatrices: Boolean = false,
) {
    CommonProfileHeader(
        userKey = user.userKey,
        bannerUrl = user.bannerUrl,
        avatarUrl = user.avatarUrl,
        displayName = user.nameElement,
        handle = user.handle,
        headerTrailing = {
            isMe.onSuccess {
                if (!it) {
                    when (relationState) {
                        is UiState.Error -> Unit
                        is UiState.Loading -> {
                            FilledTonalButton(
                                onClick = {
                                    // No-op
                                },
                                modifier =
                                    Modifier.placeholder(
                                        true,
                                        shape = ButtonDefaults.filledTonalShape,
                                    ),
                            ) {
                                Text(text = stringResource(R.string.profile_header_button_follow))
                            }
                        }

                        is UiState.Success -> {
                            when (val data = relationState.data) {
                                is UiRelation.Bluesky -> {
                                    FilledTonalButton(
                                        onClick = {
                                            onFollowClick.invoke(data)
                                        },
                                    ) {
                                        Text(
                                            text =
                                                stringResource(
                                                    when {
                                                        data.blocking -> R.string.profile_header_button_blocked
                                                        data.following -> R.string.profile_header_button_following
                                                        else -> R.string.profile_header_button_follow
                                                    },
                                                ),
                                        )
                                    }
                                }

                                else -> Unit
                            }
                        }
                    }
                }
            }
            menu.invoke(this)
        },
        content = {
            Column(
                modifier =
                    Modifier
                        .fillMaxWidth()
                        .padding(horizontal = screenHorizontalPadding),
                verticalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                user.descriptionElement?.let {
                    HtmlText2(
                        element = it,
                        layoutDirection = user.descriptionDirection,
                    )
                }
                MatricesDisplay(
                    matrices =
                        remember(user.matrices) {
                            persistentMapOf(
                                R.string.profile_header_toots_count to user.matrices.statusesCountHumanized,
                                R.string.profile_header_following_count to user.matrices.followsCountHumanized,
                                R.string.profile_header_fans_count to user.matrices.fansCountHumanized,
                            )
                        },
                    expanded = expandMatrices,
                )
            }
        },
        modifier = modifier,
        onAvatarClick = onAvatarClick,
        onBannerClick = onBannerClick,
    )
}

@Composable
internal fun ColumnScope.BlueskyUserMenu(
    user: UiUser,
    relation: UiRelation.Bluesky,
    onBlockClick: () -> Unit,
    onMuteClick: () -> Unit,
) {
    DropdownMenuItem(
        text = {
            if (relation.muting) {
                Text(text = stringResource(R.string.user_unmute, user.handle))
            } else {
                Text(text = stringResource(R.string.user_mute, user.handle))
            }
        },
        onClick = onMuteClick,
    )
    DropdownMenuItem(
        text = {
            if (relation.blocking) {
                Text(text = stringResource(R.string.user_unblock, user.handle))
            } else {
                Text(text = stringResource(R.string.user_block, user.handle))
            }
        },
        onClick = onBlockClick,
    )
}

File: iosApp/iosApp/UI/Screen/Profile/Bluesky.swift
import SwiftUI
import shared

//struct BlueskyProfileHeader: View {
//    let user: UiUser.Bluesky
//    let relation: UiState<UiRelation>
//    let isMe: UiState<KotlinBoolean>
//    let onFollowClick: (UiRelation) -> Void
//    var body: some View {
//        CommonProfileHeader(
//            bannerUrl: user.bannerUrl,
//            avatarUrl: user.avatarUrl,
//            displayName: user.extra.nameMarkdown,
//            handle: user.handle,
//            description: user.extra.descriptionMarkdown,
//            headerTrailing: { BlueskyFollowButton(relation: relation, isMe: isMe, onFollowClick: onFollowClick) },
//            content: {
//                MatrixView(
//                    followCount: user.matrices.followsCountHumanized,
//                    fansCount: user.matrices.fansCountHumanized
//                )
//            })
//    }
//}
//
//struct BlueskyFollowButton: View {
//    let relation: UiState<UiRelation>
//    let isMe: UiState<KotlinBoolean>
//    let onFollowClick: (UiRelation) -> Void
//    var body: some View {
//        if case .success(let success) = onEnum(of: isMe),
//           !success.data.boolValue,
//           case .success(let relationData) = onEnum(of: relation),
//           let actualRelation = relationData.data as? UiRelationBluesky {
//            let text = if actualRelation.blocking {
//                String(localized: "relation_blocked")
//            } else if actualRelation.following {
//                String(localized: "relation_following")
//            } else {
//                String(localized: "relation_follow")
//            }
//            Button(action: {
//                onFollowClick(actualRelation)
//            }, label: {
//                Text(text)
//            })
//            .buttonStyle(.borderless)
//        } else {
//            EmptyView()
//        }
//    }
//}
